### PR TITLE
build(common): split and parallelize `check if code is up-to-date`

### LINKED ISSUE #3100: build(common): split and parallelize `check if code is up-to-date`
## Problem

Currently the `check if code is up-to-date` step takes more than 20 minutes, and it's getting longer and longer. We can split that step into multiple steps and run them in parallel, like what we did for smoke test.

### PR DESCRIPTION
**Pull request by archerzz**
- change smoke test template into code generation test template, since:
  - the smoke test is actually a subset of coee generation test
  - both of them are run in our pipeline, which is duplication
- refactor corresponding scripts

resolve #3100

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by archerzz** [comment]
merge the code generation check into smoke test so that:
- duplication is avoided
- code generation check now runs in parallel

## Before Change

See [this job run](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2179098&view=results) which take 40 minutes.
image

## After Change

See [this job run](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2181241&view=results) which takes 17 minutes.
image

File: eng/pipelines/build.yml [hunk 1]
```diff
@@ -66,8 +66,8 @@ stages:
               DOTNET_SKIP_FIRST_TIME_EXPERIENCE: 1
               DOTNET_CLI_TELEMETRY_OPTOUT: 1
               DOTNET_MULTILEVEL_LOOKUP: 0
-          - pwsh: ./eng/CodeChecks.ps1
-            displayName: "Check if code is up-to-date"
+          - pwsh: ./eng/SharedCodeCheck.ps1
+            displayName: "Check if shared source codes are up-to-date"
```
Comment (fengzhou-msft) [inline eng/pipelines/build.yml:70]:
```suggestion
            displayName: "Check if shared source code is up-to-date"
```
I think code is uncountable when it refers to source code.

File: eng/pipelines/check-code-generation.yml [hunk 2]
```diff
@@ -19,8 +23,8 @@ jobs:
     - script: |
         npm ci
       displayName: "Install packages"
-    - pwsh: .\eng\Generate.ps1 -Filter "${{ parameters.filter }}" -Exclude TestServer,TestProjects,Samples,CadlRanchProjects -reset
-      displayName: "Check if code is up-to-date"
+    - pwsh: .\eng\CodeGenerationCheck.ps1 -Filter "${{ parameters.filter }}"
+      displayName: "Check if generated codes are up-to-date"
```
Comment (fengzhou-msft) [inline eng/pipelines/check-code-generation.yml:27]:
```suggestion
      displayName: "Check if generated code is up-to-date"
```

**Review by ArcturusZhang — APPROVED** [review]
Looks very good to me!

**Pull request merged**